custom_components/foxess_modbus/config_flow.py

```diff
@@ -108,116 +111,105 @@ def __init__(self, config=None) -> None:
+                        step_id="user",
+                        data_schema=self._modbus_type_schema,
+                    )
+                else:
+                    return await self.async_step_energy()
+            return None
+
+        return await self._with_default_form(
+            user_input, "tcp", self._modbus_tcp_schema, body
         )
 
     async def async_step_serial(self, user_input: dict[str, Any] = None):
         """Handle a flow initialized by the user."""
-        if MODBUS_SERIAL_HOST in user_input:
-            self._errors["base"] = None
-            inverter = self._parse_inverter(user_input)
-            result = await self.async_add_inverter(
-                SERIAL, user_input[MODBUS_SERIAL_HOST], inverter
-            )
-            if result and user_input[ADD_ANOTHER]:
-                self._errors["base"] = None
-                return self.async_show_form(
-                    step_id="user",
-                    data_schema=self._modbus_type_schema,
-                    errors=self._errors,
-                )
-            elif result:
-                return await self.async_step_energy()
 
-        return self.async_show_form(
-            step_id="serial",
-            data_schema=self._modbus_serial_schema,
-            errors=self._errors,
+        async def body(user_input):
+            if MODBUS_SERIAL_HOST in user_input:
+                inverter = self._parse_inverter(user_input)
+                await self.async_add_inverter(
+                    SERIAL, user_input[MODBUS_SERIAL_HOST], inverter
+                )
+                if user_input[ADD_ANOTHER]:
+                    return self.async_show_form(
+                        step_id="user",
+                        data_schema=self._modbus_type_schema,
+                    )
+                else:
+                    return await self.async_step_energy()
+            return None
+
+        return await self._with_default_form(
+            user_input, "serial", self._modbus_serial_schema, body
         )
 
     async def async_step_energy(self, user_input: dict[str, Any] = None):
         """Handle a flow initialized by the user."""
-        if user_input is not None:
+
+        async def body(user_input):
             if user_input[ENERGY_DASHBOARD]:
                 await self._setup_energy_dashboard()
             return self.async_create_entry(title=_TITLE, data=self._data)
 
-        return self.async_show_form(
-            step_id="energy",
-            data_schema=self._energy_dash,
-            errors=self._errors,
+        return await self._with_default_form(
+            user_input, "energy", self._energy_dash, body
         )
 
-    def detect_duplicate(self, inv_type, host, friendly_name):
+    def ensure_not_duplicate(self, inv_type, host, friendly_name):
         """Detect duplicates"""
         if host in self._data[inv_type]:
+            # TODO: Shouldn't we be checking for duplicate friendly names across all hosts?
```

Comment:
This looked wrong -- we allow duplicate friendly names so long as they are used with different IPs, and duplicate IPs so long as they have different friendly_names. Is that correct? I would have thought that friendly names should be unique across all connection types and addresses.

Reply:
I'm addressing this in the follow-up PR which reworks the config flow